QEP82 - User profiles

NathanW2 opened the pull request:
Super super early PR for adding User profiles based on QEP 82. Keen for feedback around design.

Adds the following new command line options:

```
qgis --profile {name}
qgis --profiles-path {path}
```

`--profile {name}` takes a name of a config to load or create.  If it's not found it will be created.
`--profiles-path {path}` . Set the root folder to look for profiles.  Will add `\profiles` to the given path.

By default profiles are loaded from `QStandardPaths::AppDataLocation` unless `QGIS_CUSTOM_CONFIG_PATH` is set, or the root path is ,

As per the QEP (which is out of date now) anything that used to be in config (registry, plist, .config, etc) and `.qgis2` is now in a single profile folder for the user profile name. These are isolated profile , more details in the QEP for why it's done this way.

A `profiles.ini` lives in the main `profiles` folder and has a pointer to the default profile (which is currently the last profile to close, unless set via the `--profile` flag.
This file will hold other options for the profiles in the future.

## Checklist

Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

NathanW2 commented:
@nyalldawson @wonder-sk @m-kuhn keen for you feedback on the design of ths.

DelazJ commented:
@NathanW2 just a recall to not forget to use [Feature] or [needs-docs] tags in your commit messages to automatically feed changelog and doc repos.
Thanks

m-kuhn commented:
Nice stuff!

What's the roaming folder exactly? Does it make sense to use it by default?

m-kuhn commented on src/app/main.cpp:
```diff
@@ -808,18 +815,51 @@ int main( int argc, char *argv[] )
+  if ( !optionpath.isEmpty() || !configpath.isEmpty() )
+  {
+    config = new QgsUserConfig( optionpath.isEmpty() ? configpath : optionpath );
+  }
```
I find it a bit hard to follow the precedence here, can this be changed to 

```
if ( !optionpath.isEmpty() )
{}
else if (!configpath.isEmpty())
{}
else if ( getenv )
{}
```

m-kuhn commented on src/core/qgsuserconfigmanager.h:
```diff
@@ -0,0 +1,33 @@
+
+#include <qgis.h>
+
+class CORE_EXPORT QgsUserConfig
```
Can you put this into its own header file, according to [the coding standards](https://www.qgis.org/en/site/getinvolved/development/qgisdevelopersguide/codingstandards.html#id1)?

m-kuhn commented on src/core/qgsuserconfigmanager.h:
```diff
@@ -0,0 +1,33 @@
+    const QString folder() const;
+
+  private:
+    QString mConfigFolder;
```
Is this expected to gain more information than the config folder? If not, could it be dumped and a simple string used instead?

NathanW2 replied:
Yeah I was thinking about expanding it out to inclulding stuff like python folder, processing folder, etc.  

Trying to avoid having people work out that path themselves.  Will move all stuff from `QgsApplication` to here if it makes sense.  

Just a thought anyway.

NathanW2 commented:
@m-kuhn about using roaming folder by default. Maybe I just need to check if it's a good idea or not first but not a big deal to me.

m-kuhn commented on src/core/qgsuserconfigmanager.h:
```diff
@@ -0,0 +1,33 @@
+#ifndef QGSUSERCONFIGMANAGER_H
+#define QGSUSERCONFIGMANAGER_H
+
+#include <qgis.h>
```
Should be `"qgis.h"`

m-kuhn commented on src/core/qgsuserconfigmanager.cpp:
```diff
@@ -0,0 +1,64 @@
+  mConfigFolder = folder;
+}
+
+const QString QgsUserConfig::folder() const
```
Should be

    QString QgsUserConfig::folder() const

m-kuhn commented on src/app/main.cpp:
```diff
@@ -808,18 +815,51 @@ int main( int argc, char *argv[] )
+    QString rootFolder = QStandardPaths::standardLocations( location ).at( 0 );
+
+    // On Windows using the roaming folder allows for folder to be uploaded to server at log off tkime.
+    QgsUserConfigManager* manager = new QgsUserConfigManager();
```
Is that ever deleted again?

m-kuhn commented:
Can you shortly explain the responsibilities of `QgsUserConfigManager` and `QgsUserConfig`?

NathanW2 commented:
Yep.  So `QgsUserConfigManager` will be used later on in the UI in order to show the list of "configs" that have been found.  I thought it was best to put into an object and attach to `QgsApplication` so that it could be used easier to work with without having to know how the configs are setup.

m-kuhn commented on src/core/qgsuserconfigmanager.cpp:
```diff
@@ -0,0 +1,64 @@
+#include <QTextStream>
+
+
+void QgsUserConfigManager::setRootFolder( QString rootConfigFolder )
```
const QString& rootConfigFolder

m-kuhn commented:
So all configs are expected to be stored under the same `rootConfigFolder` (in `QgsUserConfigManager`)? I guess that wouldn't work if you wanted a roaming and a local config side-by-side?

NathanW2 commented:
Well you can't mix the roaming and local locations in a single instance if this is what you mean.  @nyalldawson and I talked about this and it doesn't make much sense. Take this example:

```
qgis --local-config-storage
```
Will use: `C:\Users\Nathan\AppData\Local\QGIS\QGIS3\config` as the base path for all config
where as this

```
qgis
```
will use: `C:\Users\Nathan\AppData\Roaming\QGIS\QGIS3\config` as the base path for all config

I think that is the cleanest way to handle this at the moment, we can have other stuff that is stored outside of Roaming and in Local but outside if it makes sense in the future, but that isn't really related to this stuff.  Anything that is config + plugin, etc related goes into the config folder.

NathanW2 commented:
Example of my roaming folder:

![image](https://cloud.githubusercontent.com/assets/381660/23305515/e87342ac-faea-11e6-8937-285c5f649cd2.png)

I could have the same configs in Local but they are really different ones.

NathanW2 commented:
So yes you can have them side by side but not in the same running instance. Each session of QGIS is only loaded with one active config at start time.

m-kuhn commented:
I think it doesn't make any sense to mix any profiles at all (even if they are all local or all roaming)?

The `QgsUserConfigManager` should just keep track of what's available (without mixing contents) if I'm not mistaken. But in the current form it's only able to either keep track of locals or remotes within a given instance.

NathanW2 commented:
So it would be better to have QgsUserConfigManager report all configs in both local and roaming?

m-kuhn commented:
> So it would be better to have QgsUserConfigManager report all configs in both local and roaming?

You tell me ;)

I imagine within the QGIS interface it should (in the future) be possible to click "restart QGIS with config XXX" and there it would be great to have both?

NathanW2 commented:
I think generally roaming or local is a IT admin choice nothing something that the users normally cares about.  We only case about it because we are building it.

NathanW2 commented:
Let me think about it.  I will move the current QStandardPath stuff into `QgsUserConfig` and build a little UI and see how it feels.

Keen to explore how other applications do this.  Maybe Firefox does something like this.

NathanW2 commented:
I'm tempted to leave it as roaming by default and only way to get lcoal is --local-config-storage because this isn't something that most users should care about.

You might only want local storage for testing, etc and in that case keeping them apart makes sense to me for now.

m-kuhn commented:
Ideas:

 *  open it up with `--config-prefix=SomePath` (and have `--local-config-storage` a shorthand to set this to some reasonable default local folder)?
 * at some point we can add a file `configs.ini` which contains a mapping of additional config folders with absolute paths (`localConfig=C:/DevelopmentConfig; remoteConfig=Z:/CorporateConfig`).

NathanW2 commented:
I'm not sure I understand the need though we already have that.  This is really just changing the default `%HOME%/.qgis3` to `%APPDATA%\config\default`

We already have --configpath which can set the config to some place else if need if you want to have it on a network location

m-kuhn commented:
AFAICT

 * `--config` is a foldername (relative) `default` by default
 * `--local-config-storage` switches the basePath for the relative path above to something else

Why restrict the base path (from which a list of subfolders is loaded after all) to 2 locations?

Let's say I have a portable USB stick with QGIS and configurations, I could have a batch file start QGIS with `qgis.exe --config-prefix=${usb-drive}/configs` and then tell the attendees in the afternoon to switch to the profile `advanced` using the profile switcher within QGIS.

NathanW2 commented:
Ahh now I get what you are saying.  Ok yes I will make that work.  Thanks for raising that use case.

NathanW2 commented:
How how do these look

```
--config
--config-local-storage
--config-storage-prefix=
```

Just so they all line up together.

NathanW2 commented on src/app/main.cpp:
```diff
@@ -793,32 +815,65 @@ int main( int argc, char *argv[] )
+  // default config will be loaded from .default file
+  // which contains the name of the startup default config.
+
+  if ( !configLocalStorageLocation.isEmpty() )
```
@m-kuhn adding a override `--config-storage-location path` to allow for setting a custom location.

m-kuhn commented on src/app/main.cpp:
```diff
@@ -132,6 +135,9 @@ void usage( const QString &appName )
       << QStringLiteral( "\t[--dxf-preset maptheme]\tmap theme to use for dxf output\n" )
+      << QStringLiteral( "\t[--config-local-storage]\tstore user config folder in local machine folder only.\n" )
+      << QStringLiteral( "\t[--config name]\tLoad a named config from the users config folder.\n" )
+      << QStringLiteral( "\t[--config-storage-location path]\tPath to store user config folders..\n" )
```
Typo: extra-dot in the end

m-kuhn replied:
Looks like the description of the other options don't start with caps, would be good to align.
Also, I think it would be worth translating the description. Or are there technical reasons in place that don't allow this?

m-kuhn commented on src/core/qgsuserconfigmanager.h:
```diff
@@ -1,58 +1,65 @@
+     * Check if the root location has been set for the manager.
+     * @return True if the root location has been set.
+     */
+    bool rootLocationIsSet();
```
bool rootLocationIsSet() const;

m-kuhn commented on src/core/qgsuserconfigmanager.h:
```diff
@@ -1,58 +1,65 @@
+     * A list of all found config names.
+     * @return
+     */
+    QStringList allConfigNames();
```
const

m-kuhn commented on src/core/qgsuserconfigmanager.h:
```diff
@@ -1,58 +1,65 @@
+     * Returns the name of the default config that has been set in .default.
+     * @return The name of the default config.
+     */
+    QString defaultConfigName();
```
const

m-kuhn commented on src/core/qgsuserconfigmanager.h:
```diff
@@ -1,58 +1,65 @@
+     * @param name The name of the config to return.
+     * @return A QgsUserConfig pointing to the location of the user config.
+     */
+    QgsUserConfig *configForName( const QString name );
```
const

m-kuhn commented on src/core/qgsuserconfigmanager.h:
```diff
@@ -1,58 +1,65 @@
+
+    void setActiveConfig( QgsUserConfig *config );
+
+    QgsUserConfig *activeConfig();
```
const

m-kuhn commented on src/core/qgsuserconfigmanager.cpp:
```diff
@@ -1,60 +1,79 @@
+{
+  if ( !mCurrentConfig )
+  {
+    QString defaultConfig = this->defaultConfigName();
```
It's not python, no need for `this->` ;)

NathanW2 replied:
Been doing a bit too much C# today :)

nirvn commented:
@NathanW2 , in terms of terminology, I'd use user profile instead of user config.

NathanW2 commented:
Yeah, I tend to agree. It was changed in the QEP to config the more I work
on it the more Profile makes more sense.  Config is a bit confusing as it's
a lower level term vs profile which is higher level.

On Sun, Mar 12, 2017 at 12:14 AM, Mathieu Pellerin <notifications@github.com
> wrote:

> @NathanW2 <https://github.com/NathanW2> , in terms of terminology, I'd
> use user profile instead of user config.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/4184#issuecomment-285869268>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3ODN7YoqrLIRwa0x_s2a7hDgWdvsks5rkqxfgaJpZM4MKw1t>
> .
>

NathanW2 commented:
Current status update:

![config](https://cloud.githubusercontent.com/assets/381660/23882498/b6fca04a-08ac-11e7-8958-08322af53808.gif)

NathanW2 committed:
[FEATURE] Add user profile support

NathanW2 committed:
Add menu items for profiles

- New profile
- Open profile folder

NathanW2 commented:
@rduivenvoorde Updated so there is now menu items for `Open profile folder` and `New profile`.  Haven't done duplicate yet.

NathanW2 commented:
@3nids @m-kuhn what is the deal with sip now? Are all files auto generated and we just add to git?

NathanW2 committed:
Helps if we init the settings :P

DelazJ commented on src/core/qgsuserprofilemanager.h:
```diff
@@ -0,0 +1,140 @@
+    /**
+     * Return the profile from the given root profile location. If no name is given it returns a profile called "default".
+     * @param rootLocation The root location to search for the profile.
+     * @param roamingConfig True if to use Qt to find the roaming folder based on the OS. e.g APPDATA on Windows. Default true.
```
missing word or punctuation?

DelazJ commented on src/core/qgsuserprofilemanager.h:
```diff
@@ -0,0 +1,140 @@
+ *
+ * In QGIS 3 all settings, plugins, etc were moved into a %APPDATA%/profiles folder for each platform.
+ * This allows for manage different user profiles per machine vs the single default one that was allowed in the
+ * pass.
```
past

DelazJ commented on src/core/qgsuserprofilemanager.h:
```diff
@@ -0,0 +1,140 @@
+    QString defaultProfileName() const;
+
+    /**
+     * Sets the default profile name.  The default profile name is used when loading QGIS
```
spacing

DelazJ commented on src/core/qgsuserprofilemanager.h:
```diff
@@ -0,0 +1,140 @@
+    /**
+     * Create a user profile given by the name
+     * @param name
+     * @return A QgsError which report if there was any error creating the user profile.
```
reports

DelazJ reviewed (commented):
Add some typo-related comments...
Looking forward to testing

NathanW2 committed:
Fix issue of wrong root path on first run

NathanW2 committed:
Fix cli inputs

- Remove --configPath - No longer supported
- Rename --profiles-root-location to --profiles-path
- Rename --profiles-local-storage to --profiles-no-roaming (mainly for
Windows for different %APPDATA% locations)

NathanW2 committed:
Add notify on profies folder changes

NathanW2 committed:
Remove non-roaming config option

NathanW2 committed:
Add sip files

NathanW2 committed:
Add profiles folder back to root location

NathanW2 committed:
Update cli arg

NathanW2 committed:
Add tooltip for profile actions

NathanW2 committed:
API clean up

NathanW2 committed:
Debug fix up

NathanW2 committed:
Fix docs

NathanW2 committed:
More docs

NathanW2 committed:
Update sip

NathanW2 committed:
Fix sip error

NathanW2 commented on src/core/qgsuserprofile.h:
```diff
@@ -0,0 +1,76 @@
+ * User profile contains information about the user profile folders on the machine.
+ * In QGIS 3 all settings, plugins, etc were moved into a %APPDATA%/profiles folder for each platform.
+ * This allows for manage different user profiles per machine vs the single default one that was allowed in the
+ * pass.
```
Thanks.  I'll fix these up

NathanW2 commented:
@nyalldawson @m-kuhn @rduivenvoorde @nirvn this is a good state now close for merge.  Keen for any feedback before that.

I will fix the typos up in a sec.

nyalldawson commented on src/app/main.cpp:
```diff
@@ -495,6 +498,8 @@ int main( int argc, char *argv[] )
   // This behavior is used to load the app, snapshot the map,
   // save the image to disk and then exit
   QString mySnapshotFileName = QLatin1String( "" );
+  QString configLocalStorageLocation = QLatin1String( "" );
```
Not directly a result of your work - but these `QLatin1String("")`s do nothing and can be removed.

nyalldawson commented on src/app/main.cpp:
```diff
@@ -773,21 +778,30 @@ int main( int argc, char *argv[] )
   }
 
-  QgsApplication myApp( argc, argv, myUseGuiFlag, configpath );
+  // Set up the QSettings environment must be done after qapp is created
```
Is this comment redundant? It seems contradictory to the order that's happening here.

nyalldawson commented on src/app/main.cpp:
```diff
@@ -799,13 +813,7 @@ int main( int argc, char *argv[] )
-  QCoreApplication::setOrganizationDomain( QgsApplication::QGIS_ORGANIZATION_DOMAIN );
-  QCoreApplication::setApplicationName( QgsApplication::QGIS_APPLICATION_NAME );
-  QCoreApplication::setAttribute( Qt::AA_DontShowIconsInMenus, false );
+  myApp.setWindowIcon( QIcon( QgsApplication::appIconPath() ) );
```
What's happening here? looks like a duplicate of the code which is just run above.

nyalldawson commented on src/app/main.cpp:
```diff
@@ -837,25 +845,17 @@ int main( int argc, char *argv[] )
+  }
 
   // Load and set possible default customization, must be done afterQgsApplication init and QgsSettings ( QCoreApplication ) init
   QgsCustomization::instance()->setSettings( customizationsettings );
```
Not directly related to your work - but I've just noticed that customizationSettings is leaked here. (QgsCustomization is not deleting it)

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -646,6 +651,13 @@ QgisApp::QgisApp( QSplashScreen *splash, bool restorePlugins, bool skipVersionCh
   QgsRuntimeProfiler *profiler = QgsApplication::profiler();
 
+  startProfile( QStringLiteral( "User profile manager" ) );
+  mUserProfileManager = new QgsUserProfileManager();
```
mUserProfileManager is never deleted. Can you make it a unique_ptr or parent it to the app instance?

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -2230,12 +2244,56 @@ void QgisApp::createMenus()
   actionWhatsThis->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/mActionWhatsThis.svg" ) ) );
   mHelpMenu->insertAction( before, actionWhatsThis );
+
+  this->createProfileMenu();
```
redundant this

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -2230,12 +2244,56 @@ void QgisApp::createMenus()
+  menuBar()->setCornerWidget( mConfigMenuBar );
+  mConfigMenuBar->show();
+
+  this->refreshProfileMenu();
```
redundant this

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -11815,6 +11880,16 @@ void QgisApp::keyPressEvent( QKeyEvent *e )
+  if ( text.isEmpty() )
+    return;
+
+  userProfileManager()->createUserProfile( text );
```
I think the returned error should be checked after this (and feedback given to the user if something went wrong)

nyalldawson commented on src/core/qgsuserprofile.cpp:
```diff
@@ -0,0 +1,114 @@
+  QgsError error;
+  if ( !QDir( mProfileFolder ).exists() )
+  {
+    error.append( QObject::tr( "Profile folder doesn't exist" ) );
```
might be a good idea to include the missing folder name here too (helps with debugging)

nyalldawson commented on src/core/qgsuserprofile.h:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * Reference to a existing user profile folder.
+     * Profile folder should be created using QgsProfileManager.
+     * @param folder An existing profile folder as the base of the user profile.
```
\param

nyalldawson commented on src/core/qgsuserprofile.h:
```diff
@@ -0,0 +1,77 @@
+     * Profile folder should be created using QgsProfileManager.
+     * @param folder An existing profile folder as the base of the user profile.
+     */
+    QgsUserProfile( QString folder );
```
const QString &

nyalldawson commented on src/core/qgsuserprofile.h:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * Return the alias for the user profile. Reads the alias from .profile
+     * in the profile folder.
+     * @return If no alais is set name() is returned.
```
\return

nyalldawson commented on src/core/qgsuserprofile.h:
```diff
@@ -0,0 +1,77 @@
+
+    /**
+     * Set the alias of the profile.  The alias is a user friendly name.
+     * @param alias A user friendly name for the profile.
```
\param, \return

nyalldawson commented on src/core/qgsuserprofile.h:
```diff
@@ -0,0 +1,77 @@
+
+    /**
+     * The icon for the user profile.
+     * @return
```
remove empty return

nyalldawson commented on src/core/qgsuserprofilemanager.cpp:
```diff
@@ -0,0 +1,205 @@
+
+  if ( basePath.isEmpty() )
+  {
+    QString rootFolder = QStandardPaths::standardLocations( QStandardPaths::AppDataLocation ).at( 0 );
```
could this ever return an empty list? better to be safe and use .value( 0 )

nyalldawson commented on src/core/qgsuserprofilemanager.h:
```diff
@@ -0,0 +1,172 @@
+    void profilesChanged( );
+
+  private:
+    QFileSystemWatcher *mWatcher = nullptr;
```
This is leaked - use a unique_ptr instead.

nyalldawson commented on src/core/qgsuserprofilemanager.h:
```diff
@@ -0,0 +1,172 @@
+
+    QString mRootProfilePath;
+
+    QgsUserProfile *mUserProfile = nullptr;
```
This is also leaked a number of times - switch it to a unique_ptr

nyalldawson commented on src/core/qgsuserprofilemanager.h:
```diff
@@ -0,0 +1,172 @@
+
+    QString settingsFile();
+
+    QSettings *mSettings = nullptr;
```
Also leaked - flip to unique_ptr

nyalldawson commented on src/core/qgsuserprofilemanager.h:
```diff
@@ -0,0 +1,172 @@
+     * @param name The name of the profile to return.
+     * @return A QgsUserprofile pointing to the location of the user profile.
+     */
+    QgsUserProfile *profileForName( const QString name ) const;
```
needs SIP_FACTORY annotation (and better note in the docs that it returns a new profile, caller responsible for cleanup)

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -2230,12 +2244,56 @@ void QgisApp::createMenus()
+
+  Q_FOREACH ( const QString &name, userProfileManager()->allProfiles() )
+  {
+    profile = userProfileManager()->profileForName( name );
```
This is a bit messy - profileForName returns a new object, which is getting leaked here, vs the earlier assigned value of userProfile() which doesn't transfer ownership

nyalldawson commented on src/core/qgsuserprofilemanager.cpp:
```diff
@@ -0,0 +1,205 @@
+void QgsUserProfileManager::loadUserProfile( const QString &name )
+{
+  QgsUserProfile *profile = profileForName( name );
+  loadUserProfile( profile );
```
profile is leaked here

nyalldawson commented on src/core/qgsuserprofilemanager.h:
```diff
@@ -0,0 +1,172 @@
+     * @param createNew Create the profile folder if it doesn't exist.
+     * @return The user profile
+     */
+    QgsUserProfile *getProfile( const QString &defaultProfile = "default", bool createNew = true, bool initSettings = true );
```
Needs SIP_FACTORY, and some note in docs that a new object is returned

nyalldawson commented on src/app/main.cpp:
```diff
@@ -773,21 +778,30 @@ int main( int argc, char *argv[] )
+
+  QString rootProfileFolder = QgsUserProfileManager::resolveProfilesFolder( configLocalStorageLocation );
+  QgsUserProfileManager manager( rootProfileFolder );
+  QgsUserProfile *profile = manager.getProfile( profileName, true );
```
profile is leaked here

nyalldawson commented:
Hey @NathanW2 - just done a bit of review on the code above. All looks good except for a bunch of leaks which need plugging. I haven't reviewed the logic here - I think @elpaso or @dakcarto could review that?

NathanW2 committed:
Updates from comments

elpaso commented on src/core/qgsuserprofilemanager.cpp:
```diff
@@ -0,0 +1,200 @@
+  // http://doc.qt.io/qt-5/qcoreapplication.html#arguments
+  arguments.removeFirst();
+
+  arguments << "--profile" << name;
```
I've not tested, but what happens here if qgis was already launched with `--profile` ? Does the second override the first? If this is not the case, a more robust approach would be required to parse the arguments.

NathanW2 replied:
Untested but will test now.  I suspect the last will win.

NathanW2 replied:
Confirmed the last one wins. So it's not clean but I'm ok with this for now.

elpaso commented on src/core/qgsuserprofilemanager.cpp:
```diff
@@ -0,0 +1,200 @@
+
+QString QgsUserProfileManager::resolveProfilesFolder( const QString &basePath )
+{
+  if ( getenv( "QGIS_CUSTOM_CONFIG_PATH" ) )
```
As suggested in the general comments, I'd prefer if the `getenv` would be moved to main.cpp and its value passed as an argument to the constructor as an input to this class.

NathanW2 replied:
Will do I agree it makes more sense.

elpaso reviewed (commented):
Many many thanks for this great new feature!!!

## General observations about initial defaults

I've added some comments about the possibility to also get the `profile` and `profiles-path` from the global settings (QgsSettings fallback) as a first run default, that could be useful for some complex enterprise deployments where the sysadmin wants to provide a pre-configured profile in user space as part of the deployment, in this case, the sysadmin should be able to provide the `profile` and `profile-path` initial defaults in the global settings file.

Also, I would suggest to move all the logic to collect the profile and path configuration from the environment outside of the classes and into the main.cpp, like it is already done for other input arguments.


## Possible bug ?

While testing on Linux, I've not found a way to "remember" the profile from one session to another, here is my (possibly wrong) testing procedure:

- launch qgis instance 1 from the command line with no arguments
- create a new profile "myprofile"
- switch to the new profile (instance 2 opens)
- quit instance 1
- quit instance 2
- re-launch qgis from the command line with no arguments

expected behavior:
- qgis starts with the new profile "myprofile" activated since that was the last profile used (by instance 2)
actual behavior:
- qgis starts with the "default" profile

NathanW2 committed:
Updates from comments

NathanW2 committed:
Fix missing path

NathanW2 commented:
@elpaso I have merged this so it doesn't keep sitting here but I will address you requirements for loading the name from global in master

NathanW2 commented:
@elpaso changes to load from global are now in master.

NathanW2 commented:
Merged as 7b92f1f

jgrocha commented:
Hi @NathanW2,
Great work with profiles! I'm testing it. There is an empty menu entry for each item in my linux version. Is this expected?

![empty menu entries profiles](https://user-images.githubusercontent.com/163681/28939293-ba051460-7888-11e7-83be-6f8bd457fee4.png)

rduivenvoorde commented:
I think that is a (Ubuntu) theming issue, this is how it looks with me:

![selection_056](https://user-images.githubusercontent.com/731673/28941020-e68ca198-7896-11e7-9b92-9d79d0eb84b7.png)

I would be ok to just style it normal, the grey/white styling (or at least in this style/theme) there I do not see anywhere else in QGIS?

jgrocha commented:
Thanks @rduivenvoorde 

It is definitely a Ubuntu problem. I don't see any other usage of menu section on QGIS. That's why I never missed menu entries before.

I'll need to investigate this issue further.
